source/src/BNFC/Backend/FSharp.hs

```diff
@@ -0,0 +1,246 @@
+    mkfile (printerFile opts)   comment $ cf2Printer prMod absMod cf
+    mkfile (showFile opts)      comment $ cf2show showMod absMod cf
+    mkfile (tFile opts)         comment $ fsharpTestfile absMod lexMod parMod prMod showMod tFileMod cf
+    mkfile (utilFile opts)      comment $ utilM (utilFileM opts)
+    mkfile (fsprojFile opts)    XML.comment $ fsprojM opts
+    mkMakefile opts $ makefile opts
+    -- case xml opts of
+    --   2 -> makeXML opts True cf
+    --   1 -> makeXML opts False cf
+    --   _ -> return ()
+
+-- | Generate the makefile.
+makefile
+  :: SharedOptions
+  -> String    -- ^ Filename of the makefile.
+  -> Doc       -- ^ Content of the makefile.
+makefile opts makeFile = vcat
+  [ "# Makefile for building the parser and test program."
+  , phonyRule
+  , defaultRule
+  , vcat [ "# Rules for building the parser." , "" ]
+  -- If option -o was given, we have no access to the grammar file
+  -- from the Makefile.  Thus, we have to drop the rule for
+  -- reinvokation of bnfc.
+  , when (isDefault outDir opts) $ bnfcRule
+  , testParserRule
+  , vcat [ "# Rules for cleaning generated files." , "" ]
+  , cleanRule
+  , distCleanRule
+  , "# EOF"
+  ]
+  where
+  -- | List non-file targets here.
+  phonyRule :: Doc
+  phonyRule = vcat
+    [ "# List of goals not corresponding to file names."
+    , ""
+    , Makefile.mkRule ".PHONY" [ "all", "clean", "distclean" ] []
+    ]
+  -- | Default: build test parser(s).
+  defaultRule :: Doc
+  defaultRule = vcat
+     [ "# Default goal."
+     , ""
+     , Makefile.mkRule "all" tgts []
+     ]
+     where
+     tgts = [ buildTarget opts ]
+
+  -- | Rule to build F# test parser.
+  testParserRule :: Doc
+  testParserRule = Makefile.mkRule tgt deps [ "dotnet build" ]
+    where
+    tgt :: String
+    tgt = buildTarget opts
+    deps :: [String]
+    deps = map ($ opts)
+      [   absFile
+        , printerFile
+        , tFile
+        , fslexFile
+        , fsyaccFile
+        , templateFile
+        , showFile
+        , utilFile
+        , fsprojFile
+      ]
+  cleanRule =
+      mkRule "clean" []
+          [ "-rm -fr bin obj "]
+
+  distCleanRule =
+       mkRule "distclean" ["clean"]
+          [ "-rm -f " ++ unwords [ mkFile withLang "Lex" "*" opts,
+                                   mkFile withLang "Par" "*" opts,
+                                   mkFile withLang "Layout" "*" opts,
+                                   mkFile withLang "Skel" "*" opts,
+                                   mkFile withLang "Print" "*" opts,
+                                   mkFile withLang "Show" "*" opts,
+                                   mkFile withLang "Test" "*" opts,
+                                   mkFile withLang "Abs" "*" opts,
+                                   mkFile withLang "Test" "" opts,
+                                   mkFile withLang  "" "fsproj" opts,
+                                   utilFile opts,
+                                   makeFile ]]
```

Comment:
I used the code from Haskell as a reference for this part, but I don't understand why it removes the makefile.

Reply:
This is indeed debatable.  I think the rationale was that `distclean` should remove all generated files, and  the `Makefile` was one of them.  However, this could be dangerous if the user edited the `Makefile` later.  Atm, the user also has to edit the `distclean` goal then.
Personally, I do not use the `distclean`, but there is a test for it in `bnfc-system-tests`.  Thus, I would just keep it like this.  
A smarter cleaning logic would be another project: store hashes along generated files and only automatically delete stuff that hasn't changed since generated last (checkable using the hash)...